### PR TITLE
Silencing numpy warnings

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] Closes https://github.com/pydata/xarray/issues/3811
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

Is this the right approach? Or should we be using `np.errstate` machinery?

### CODE REVIEW

**Comment by mathause** [comment]
I think this thread suggests to use `warning`: https://github.com/numpy/numpy/issues/10709

File: xarray/tests/test_dataset.py [hunk 1]
```diff
@@ -6191,6 +6191,33 @@ def test_raise_no_warning_for_nan_in_binary_ops():
...
+    with pytest.warns(None) as record:
+        assert_allclose(ds, ds)
+
+    assert len(record) == 0
+
+
+@pytest.mark.xfail()
+@pytest.mark.parametrize("ds", (2,), indirect=True)
+@pytest.mark.parametrize("name", ("mean", "max"))
+def test_raise_no_warning_dask_rolling_assert_close(ds, name):
```
Comment (max-sixty) [inline xarray/tests/test_dataset.py:6206]: This is a puzzle — I can't easily find the source of the warning. It requires `assert_allclose` to be run, for the `ds` param to be 2, and is different for `mean` and `max`. `sum` raises no warning.

Unless we can easily figure this out, I'd be up for xfailing, leaving a note, and merging this — it does seem to be a corner case.

File: xarray/tests/test_dataset.py [hunk 2]
```diff
@@ -6191,6 +6191,41 @@ def test_raise_no_warning_for_nan_in_binary_ops():
...
+    ds = ds.chunk({"x": 4})
+
+    with pytest.warns(None) as record:
```
Comment (keewis) [inline xarray/tests/test_dataset.py:6209]: If I remember correctly, we can just decorate the test with `pytest.mark.filterwarnings("error")` to convert warnings to errors, which additionally preserves the traceback.

File: xarray/tests/test_dataset.py [hunk 3]
```diff
@@ -6191,6 +6191,41 @@ def test_raise_no_warning_for_nan_in_binary_ops():
...
+
+        rolling_obj = ds.rolling(
+            time=4,
+            x=3,
+        )
```
Comment (keewis) [inline xarray/tests/test_dataset.py:6214]: if you remove the trailing comma, `black` will collapse this back into a single line

File: xarray/tests/test_dataset.py (hunk 2 above)
Comment (max-sixty) [reply xarray/tests/test_dataset.py]: Excellent, thank you @keewis !

File: xarray/tests/test_dataset.py [hunk 4]
```diff
@@ -6191,6 +6191,34 @@ def test_raise_no_warning_for_nan_in_binary_ops():
...
+
+    rolling_obj = ds.rolling(time=4, x=3)
+
+    actual = getattr(rolling_obj, name)()
+    expected = getattr(
+        getattr(ds.rolling(time=4,), name)().rolling(
+            x=3,
+        ),
+        name,
+    )()
```
Comment (keewis) [inline xarray/tests/test_dataset.py:6218]: there are more spurious trailing commas here:
```suggestion
    expected = getattr(getattr(ds.rolling(time=4), name)().rolling(x=3), name)()
```

File: xarray/tests/test_dataset.py (hunk 1 above)
Comment (keewis) [reply xarray/tests/test_dataset.py:6207]: can you add a reason to the xfail? That way understanding the test might be a little easier.

File: xarray/tests/test_dataset.py (hunk 4 above)
Comment (max-sixty) [reply xarray/tests/test_dataset.py]: Thanks!

File: xarray/core/dataarray.py [hunk 5]
```diff
@@ -2709,8 +2708,13 @@ def __rmatmul__(self, other):
     def _unary_op(f: Callable[..., Any]) -> Callable[..., "DataArray"]:
         @functools.wraps(f)
         def func(self, *args, **kwargs):
-            with np.errstate(all="ignore"):
-                return self.__array_wrap__(f(self.variable.data, *args, **kwargs))
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", r"All-NaN (slice|axis) encountered")
+                warnings.filterwarnings(
+                    "ignore", r"Mean of emmty slice", category=RuntimeWarning
```
Comment (max-sixty) [inline xarray/core/dataarray.py:2714]:
```suggestion
                    "ignore", r"Mean of empty slice", category=RuntimeWarning
```

**Review by keewis — APPROVED** [review]
LGTM, but the linter-enforced else clause seems a bit strange to me

**Comment by max-sixty** [comment]
> LGTM, but the linter-enforced else clause seems a bit strange to me

Yeah, I think it makes sense the context of the function alone — it doesn't know that the values are limited to the two parameterizations. Though this fix trades one error for another, in the end.
